File: Pacode/2026.2/turma_intermediaria/ListaIntermediaria_CoordinateCompression_SparseTable/F - Sereja and D/F-sereja-and-d.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> values(n);
    for (int &x : values) cin >> x;

    int m;
    cin >> m;

    vector<vector<int>> table;
    vector<int> gaps;
    for (int i = 0; i + 1 < n; i++) gaps.push_back(values[i + 1] - values[i]);
    table.push_back(gaps);

    for (int k = 1; (1 << k) <= n - 1; k++) {
        int half = 1 << (k - 1);
        int length = n - (1 << k);
        vector<int> row(length);
        for (int i = 0; i < length; i++) row[i] = max(table[k - 1][i], table[k - 1][i + half]);
        table.push_back(row);
    }

    while (m--) {
        int t, d;
        cin >> t >> d;
        int pos = upper_bound(values.begin(), values.end(), t) - values.begin() - 1;

        for (int k = (int)table.size() - 1; k >= 0; k--) {
            int jump = 1 << k;
            if (pos >= jump && table[k][pos - jump] <= d) pos -= jump;
        }

        cout << pos + 1 << '\n';
    }
}

File: Pacode/2026.2/turma_intermediaria/ListaIntermediaria_CoordinateCompression_SparseTable/F - Sereja and D/F-sereja-and-d.py
import sys
from bisect import bisect_right

data = iter(map(int, sys.stdin.buffer.read().split()))
n = next(data)
values = [next(data) for _ in range(n)]
m = next(data)

gaps = [values[i + 1] - values[i] for i in range(n - 1)]
table = [gaps]
k = 1
while 1 << k <= n - 1:
    previous = table[-1]
    half = 1 << (k - 1)
    length = n - (1 << k)
    table.append([
        previous[i] if previous[i] > previous[i + half] else previous[i + half]
        for i in range(length)
    ])
    k += 1

answer = []
for _ in range(m):
    t = next(data)
    d = next(data)
    pos = bisect_right(values, t) - 1

    for k in range(len(table) - 1, -1, -1):
        jump = 1 << k
        if pos >= jump and table[k][pos - jump] <= d:
            pos -= jump

    answer.append(str(pos + 1))

sys.stdout.write('\n'.join(answer))
